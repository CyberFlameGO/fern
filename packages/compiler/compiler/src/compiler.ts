import { SyntaxAnalysisStage } from "@fern/syntax-analysis";
import { FernFile } from "@usebirch/compiler-commons";
import { IntermediateRepresentationGenerationStage } from "@usebirch/ir-generation";
import { Compiler, CompilerFailureType } from "./types";

export async function compile(files: readonly FernFile[]): Promise<Compiler.Result> {
    const syntaxAnalysisResult = await SyntaxAnalysisStage.run(files);
    if (!syntaxAnalysisResult.didSucceed) {
        return {
            didSucceed: false,
            failure: {
                type: CompilerFailureType.SYNTAX_ANALYSIS,
                failures: syntaxAnalysisResult.failures,
            },
        };
    }

    const intermediateRepresentationResult = await IntermediateRepresentationGenerationStage.run(
        syntaxAnalysisResult.result.validatedFiles
    );
    if (!intermediateRepresentationResult.didSucceed) {
        return {
            didSucceed: false,
            failure: {
                type: CompilerFailureType.IR_GENERATION,
            },
        };
    }

    return {
        didSucceed: true,
        intermediateRepresentation: intermediateRepresentationResult.result,
    };
}
