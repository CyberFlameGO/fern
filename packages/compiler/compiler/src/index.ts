export { compile } from "./compiler";
export type { Compiler } from "./types";
export { CompilerFailureType } from "./types";
