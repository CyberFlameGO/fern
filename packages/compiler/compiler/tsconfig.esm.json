{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/esm", "rootDir": "src", "module": "esnext" },
  "include": ["./src"],
  "references": [
    { "path": "../commons/tsconfig.esm.json" },
    { "path": "../ir-generation/tsconfig.esm.json" },
    { "path": "../syntax-analysis/tsconfig.esm.json" }
  ]
}
