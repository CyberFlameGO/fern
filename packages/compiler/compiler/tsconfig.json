{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "dist/esm", "rootDir": "src" },
  "include": ["./src"],
  "references": [
    { "path": "../commons/tsconfig.json" },
    { "path": "../ir-generation/tsconfig.json" },
    { "path": "../ir-types/fern-ir-typescript/tsconfig.json" },
    { "path": "../syntax-analysis/tsconfig.json" }
  ]
}
