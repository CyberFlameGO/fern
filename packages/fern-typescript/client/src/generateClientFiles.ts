import { IntermediateRepresentation } from "@fern-api/api";
import { TypeResolver, withDirectory } from "@fern-typescript/commons";
import { generateErrorFiles } from "@fern-typescript/errors";
import { generateModelFiles } from "@fern-typescript/model";
import { Directory } from "ts-morph";
import { generateHttpService } from "./generateHttpService";

export function generateClientFiles({
    intermediateRepresentation,
    directory,
}: {
    intermediateRepresentation: IntermediateRepresentation;
    directory: Directory;
}): void {
    const typeResolver = new TypeResolver(intermediateRepresentation);

    const modelDirectory = generateModelFiles({
        directory,
        intermediateRepresentation,
        typeResolver,
    });

    const errorsDirectory = generateErrorFiles({
        directory,
        intermediateRepresentation,
        modelDirectory,
    });

    withDirectory({ containingModule: directory, name: "services" }, (servicesDirectory) => {
        for (const service of intermediateRepresentation.services.http) {
            generateHttpService({
                service,
                servicesDirectory,
                modelDirectory,
                errorsDirectory,
                typeResolver,
            });
        }
    });
}
