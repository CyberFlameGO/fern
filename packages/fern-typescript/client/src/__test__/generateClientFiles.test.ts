import { compile } from "@fern-api/compiler";
import { writeFiles } from "@fern-api/typescript-commons";
import { parseFernDirectory } from "fern-api";
import { rm } from "fs/promises";
import { vol } from "memfs";
import path from "path";
import { Project } from "ts-morph";
import { generateClientFiles } from "../generateClientFiles";

const MOCK_APIS_DIR = path.join(__dirname, "mocks");
const GENERATED_DIR = path.join(__dirname, "generated");

beforeAll(async () => {
    await rm(GENERATED_DIR, { recursive: true, force: true });
});

describe("generateClientFiles", () => {
    it("posts", async () => {
        const directory = path.join(MOCK_APIS_DIR, "posts");
        const generatedDir = path.join(directory, "generated");

        const files = await parseFernDirectory(path.join(directory, "src"));
        const compilerResult = await compile(files);
        if (!compilerResult.didSucceed) {
            throw new Error(JSON.stringify(compilerResult.failure));
        }

        const project = new Project({
            useInMemoryFileSystem: true,
        });

        generateClientFiles({
            directory: project.createDirectory("src"),
            intermediateRepresentation: compilerResult.intermediateRepresentation,
        });

        await writeFiles("/", project, vol.promises);
        expect(vol.toJSON()).toMatchSnapshot();

        await writeFiles(generatedDir, project);
    }, 15_000);
});
