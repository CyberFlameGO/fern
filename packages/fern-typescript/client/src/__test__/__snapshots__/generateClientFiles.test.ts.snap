// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateClientFiles posts 1`] = `
Object {
  "/src/errors/index.ts": "export * from \\"./posts\\";
",
  "/src/errors/posts/PostNotFoundError.ts": "import * as model from \\"../../model\\";

export interface PostNotFoundError {
    id: model.PostId;
}
",
  "/src/errors/posts/index.ts": "export * from \\"./PostNotFoundError\\";
",
  "/src/index.ts": "export * from \\"./errors\\";
export * from \\"./model\\";
export * from \\"./services\\";
",
  "/src/model/commons/PostId.ts": "/**
 * An ID of a blog post
 */
export type PostId = string & {
    __PostId: void,
};

export const PostId = {
    of: (value: string): PostId => value as PostId
};
",
  "/src/model/commons/index.ts": "export * from \\"./PostId\\";
",
  "/src/model/index.ts": "export * from \\"./commons\\";
export * from \\"./posts\\";
",
  "/src/model/posts/Author.ts": "export type Author =
    | Author.Anonymous
    | Author.Named;

export declare namespace Author {
    interface Anonymous {
        _type: \\"anonymous\\";
    }

    interface Named {
        _type: \\"named\\";
        named: string;
    }

    export interface _Visitor<Result> {
        anonymous: () => Result;
        named: (value: string) => Result;
        _unknown: () => Result;
    }
}

export const Author = {
    anonymous: (): Author.Anonymous => ({
        _type: \\"anonymous\\"
    }),

    named: (value: string): Author.Named => ({
        named: value,
        _type: \\"named\\"
    }),

    _visit: <Result>(value: Author, visitor: Author._Visitor<Result>): Result => {
        switch (value._type) {
            case \\"anonymous\\": return visitor.anonymous();
            case \\"named\\": return visitor.named(value.named);
            default: return visitor._unknown();
        }
    },
};
",
  "/src/model/posts/Post.ts": "import { PostId } from \\"../commons/PostId\\";
import { Author } from \\"./Author\\";
import { PostType } from \\"./PostType\\";

/**
 * A blog post
 */
export interface Post {
    id: PostId;
    type: PostType;
    title: string;
    author: Author;
    content: string;
}
",
  "/src/model/posts/PostType.ts": "export type PostType =
    | PostType.Long
    | PostType.Short;

export const PostType = {
    Long: \\"LONG\\" as PostType.Long,
    Short: \\"SHORT\\" as PostType.Short,

    _visit: <Result>(value: PostType, visitor: PostType._Visitor<Result>): Result => {
        switch (value) {
            case PostType.Long: return visitor.long();
            case PostType.Short: return visitor.short();
            default: return visitor._unknown();
        }
    },
};

export declare namespace PostType {
    export type Long = \\"LONG\\" & {
        __Long: void,
    };
    export type Short = \\"SHORT\\" & {
        __Short: void,
    };

    export interface _Visitor<Result> {
        long: () => Result;
        short: () => Result;
        _unknown: () => Result;
    }
}
",
  "/src/model/posts/index.ts": "export * from \\"./Author\\";
export * from \\"./Post\\";
export * from \\"./PostType\\";
",
  "/src/services/PostsService/PostsService.ts": "import { defaultFetcher, Fetcher, joinPaths, Service } from \\"@fern-typescript/service-utils\\";
import * as endpoints from \\"./endpoints\\";

export interface Client {
    createPost(request: endpoints.createPost.RequestBody): Promise<endpoints.createPost.Response>;
    getPost(request: endpoints.getPost.Request): Promise<endpoints.getPost.Response>;
}

export class Client implements Client {
    private baseUrl: string;
    private fetcher: Fetcher;

    constructor(args: Service.Init) {
        this.fetcher = args.fetcher ?? defaultFetcher;
        this.baseUrl = joinPaths(args.serverUrl, args.serviceBaseUrl);
    }

    public async createPost(request: endpoints.createPost.RequestBody): Promise<endpoints.createPost.Response> {
        const response = await this.fetcher({
            url: joinPaths(this.baseUrl, \\"/\\"),
            method: \\"POST\\",
            headers: {},
            body: request
        });

        if (response.statusCode >= 200 && response.statusCode < 300) {
            return {
                ok: true,
                statusCode: response.statusCode,
                body: response.body as endpoints.createPost.ResponseBody
            };
        }
        else {
            return {
                ok: false,
                statusCode: response.statusCode,
                error: response.body as endpoints.createPost.ErrorBody
            };
        }
    }

    public async getPost(request: endpoints.getPost.Request): Promise<endpoints.getPost.Response> {
        const queryParameters = new URLSearchParams();
        if (request.page != null) {
            queryParameters.append(\\"page\\", request.page.toString());
        }
        queryParameters.append(\\"otherParam\\", request.otherParam.toString());

        const response = await this.fetcher({
            url: joinPaths(this.baseUrl, \`/\${request.postId}\`),
            method: \\"GET\\",
            headers: {},
            queryParameters
        });

        if (response.statusCode >= 200 && response.statusCode < 300) {
            return {
                ok: true,
                statusCode: response.statusCode
            };
        }
        else {
            return {
                ok: false,
                statusCode: response.statusCode,
                error: response.body as endpoints.getPost.ErrorBody
            };
        }
    }
}
",
  "/src/services/PostsService/endpoints/createPost/ErrorBody.ts": "export type ErrorBody = never;

export declare namespace ErrorBody {
    export interface _Visitor<Result> {
        _unknown: () => Result;
    }
}

export const ErrorBody = {
    _visit: <Result>(_value: ErrorBody, visitor: ErrorBody._Visitor<Result>): Result => {
        return visitor._unknown();
    },
};
",
  "/src/services/PostsService/endpoints/createPost/RequestBody.ts": "import * as model from \\"../../../../model\\";

export interface RequestBody {
    type: model.PostType;
    title: string;
    author: model.Author;
    content: string;
}
",
  "/src/services/PostsService/endpoints/createPost/Response.ts": "import { ErrorBody } from \\"./ErrorBody\\";
import { ResponseBody } from \\"./ResponseBody\\";

export type Response = SuccessResponse | ErrorResponse;

export interface SuccessResponse {
    ok: true;
    statusCode: number;
    body: ResponseBody;
}

export interface ErrorResponse {
    ok: false;
    statusCode: number;
    error: ErrorBody;
}
",
  "/src/services/PostsService/endpoints/createPost/ResponseBody.ts": "import * as model from \\"../../../../model\\";

export interface ResponseBody {
    type: model.PostType;
    title: string;
    author: model.Author;
    content: string;
}
",
  "/src/services/PostsService/endpoints/createPost/index.ts": "export * from \\"./ErrorBody\\";
export * from \\"./RequestBody\\";
export * from \\"./Response\\";
export * from \\"./ResponseBody\\";
",
  "/src/services/PostsService/endpoints/getPost/ErrorBody.ts": "import * as model from \\"../../../../errors\\";

export type ErrorBody =
    | ErrorBody.NotFound;

export declare namespace ErrorBody {
    interface NotFound extends model.PostNotFoundError {
        _type: \\"notFound\\";
    }

    export interface _Visitor<Result> {
        notFound: (value: model.PostNotFoundError) => Result;
        _unknown: () => Result;
    }
}

export const ErrorBody = {
    notFound: (value: model.PostNotFoundError): ErrorBody.NotFound => ({
        ...value,
        _type: \\"notFound\\"
    }),

    _visit: <Result>(value: ErrorBody, visitor: ErrorBody._Visitor<Result>): Result => {
        switch (value._type) {
            case \\"notFound\\": return visitor.notFound(value);
            default: return visitor._unknown();
        }
    },
};
",
  "/src/services/PostsService/endpoints/getPost/Request.ts": "import * as model from \\"../../../../model\\";

export interface Request {
    postId: model.PostId;
    page: number | null | undefined;
    otherParam: number;
}
",
  "/src/services/PostsService/endpoints/getPost/Response.ts": "import { ErrorBody } from \\"./ErrorBody\\";

export type Response = SuccessResponse | ErrorResponse;

export interface SuccessResponse {
    ok: true;
    statusCode: number;
}

export interface ErrorResponse {
    ok: false;
    statusCode: number;
    error: ErrorBody;
}
",
  "/src/services/PostsService/endpoints/getPost/index.ts": "export * from \\"./ErrorBody\\";
export * from \\"./Request\\";
export * from \\"./Response\\";
",
  "/src/services/PostsService/endpoints/index.ts": "export * as createPost from \\"./createPost\\";
export * as getPost from \\"./getPost\\";
",
  "/src/services/PostsService/index.ts": "export * as Endpoints from \\"./endpoints\\";
export * from \\"./PostsService\\";
",
  "/src/services/index.ts": "export * as PostsService from \\"./PostsService\\";
",
}
`;
